### PR TITLE
Sar download endpoint

### PR DESCRIPTION
**Pull request by Calum-Campbell**
- Endpoint for requesting Subject Access Request files
- Returns .txt file

Related user-admin PR: https://github.com/guardian/identity-admin/pull/245

### CODE REVIEW

File: app/controllers/UsersController.scala [hunk 1]
```diff
@@ -231,4 +232,21 @@ import models.client.ApiError._
...
+  def userSubjectAccessRequest(id: String) = auth.async { implicit request =>
+    logger.info(s"Getting subject access request for user with id: $id")
+    withMetricsF("userSubjectAccessRequest") {
+      val userSarResponse = userService.getSubjectAccessRequest(id)
+
+      userSarResponse.map {
+        case \/-(result) => {
+
+          val fileToSend = Files.createTempFile(s"sar-$id", ".txt").toFile
+          new PrintWriter(fileToSend) { result.foreach(write) }.close()
```
Comment (QuarpT) [inline app/controllers/UsersController.scala]: it's a bit unusual to call result.foreach(write) inside the constructor

File: app/controllers/UsersController.scala [hunk 2]
```diff
@@ -231,4 +232,21 @@ import models.client.ApiError._
...
+
+  def userSubjectAccessRequest(id: String) = auth.async { implicit request =>
+    logger.info(s"Getting subject access request for user with id: $id")
+    withMetricsF("userSubjectAccessRequest") {
+      val userSarResponse = userService.getSubjectAccessRequest(id)
+
+      userSarResponse.map {
+        case \/-(result) => {
+
+          val fileToSend = Files.createTempFile(s"sar-$id", ".txt").toFile
```
Comment (QuarpT) [inline app/controllers/UsersController.scala]: is it possible to do this in memory without a temporary file?

File: app/controllers/UsersController.scala (hunk 1 above)
Comment (Calum-Campbell) [reply app/controllers/UsersController.scala]:
```
      userSarResponse.map {
        case \/-(result) => {
          val fileToSend = Files.createTempFile(s"sar-$id", ".txt").toFile
          val pw = new PrintWriter(fileToSend)
          result.foreach(pw.write)
          pw.close()
          Ok.sendFile(fileToSend, onClose = () => { fileToSend.delete() })
        }
        case -\/(error) => InternalServerError(error)
      }
```

File: conf/routes [hunk 3]
```diff
@@ -2,6 +2,7 @@ GET        /healthcheck                             @controllers.HealthCheckCont
 GET        /authHealthcheck                         @controllers.HealthCheckController.authHealthCheck
 
 GET        /v1/user/search                          @controllers.UsersController.search(query: String, limit: Option[Int], offset: Option[Int])
+            GET        /v1/user/:id/sar             @controllers.UsersController.userSubjectAccessRequest(id: String)
```
Comment (rahilb) [inline conf/routes]: formatting is strange 👾

File: app/models/database/postgres/PostgresSubjectAccessRequestRepository.scala [hunk 4]
```diff
@@ -0,0 +1,113 @@
+package models.database.postgres
```
Comment (QuarpT) [inline app/models/database/postgres/PostgresSubjectAccessRequestRepository.scala]: This needs unit tests

File: app/models/database/postgres/PostgresSubjectAccessRequestRepository.scala [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
```
Comment (QuarpT) [inline app/models/database/postgres/PostgresSubjectAccessRequestRepository.scala]: The pattern followed has been one `Postgres*Repository.scala` file for each repo
You could implement each `get` dump in the `Postgres*Repository.scala` files and create a new SarService which does the combining `subjectAccessRequestById`

**Pull request merged**